### ISSUE TITLE
Add support for index buffers

### ISSUE THREAD

**Issue by fatcerberus** [spheredev/neosphere#193]
A `Shape` currently encapsulates a vertex buffer and a texture.  This works well for simple shapes, but more complex ones would benefit from deduplication via an index buffer.

Add the ability to make an "indexed shape", where the shape's vertices are simply integer offsets into a separate vertex buffer.  This may require the introduction of new objects to the API.

**Comment by fatcerberus** [comment]
This is now supported internally.  There are a few limitations, namely that vertex and index buffers are maintained internally by the Shape object and can't be swapped out on-the-fly or reused in any way.  That can be done, but would require refactoring the way the Galileo API is structured, and I rather like its current simplicity.

**Comment by fatcerberus** [comment]
In the process of implementing this and fixing issues I found along the way, I uncovered a long-standing bug: After a few hundred shapes the engine seems to give up and starts failing to create vertex buffers.  This went unnoticed for a long time for two reasons: 1) The `Shape` constructor didn't check for failure to upload, and 2) There was an internal fallback for old GPUs without vertex buffer support.  Now that the check is in place and the fallback no longer exists, the issue has reared its ugly head.

...too bad I don't know what's wrong or how to fix it yet.

**Comment by fatcerberus** [comment]
The failing VBO creation turned out to be a bug in Allegro where it wasn’t clearing the error state from an earlier failure, causing it to be picked up during VBO a creation.  Worked around it by clearing the error manually beforehand.

I worked more on internal support for independent buffers, and the code is at a point now where I’m confident I can get the necessary API changes into 4.7.  So stay tuned for awesomeness. :o)

**Comment by fatcerberus** [comment]
Implemented in master.  This ended up being a much more sweeping change than originally planned, introducing two entirely new objects to the API: `VertexList` and `IndexList`.  That allows a shape's vertices to be replaced after creation, which wasn't previously possible, as well as letting multiple shapes pull from a common vertex buffer.

While I did imply above I didn't want to change `Shape` for fear of losing Galileo's simplicity, it seems my fears were unfounded.  Any increase in complexity (which is minimal) is *more* than offset by the additional flexibility the new objects bring.
